les/lespay/server/balance.go

```diff
@@ -0,0 +1,803 @@
+
+var errBalanceOverflow = errors.New("balance overflow")
+
+const (
+	maxBalance               = math.MaxInt64   // maximum allowed balance value
+	posThreshold             = 1000000         // minimum positive balance that is persisted in the database
+	negThreshold             = 1000000         // minimum negative balance that is persisted in the database
+	persistExpirationRefresh = time.Minute * 5 // refresh period of the token expiration persistence
+)
+
+const (
+	balanceCallbackUpdate = iota // called when priority drops below the last minimum estimate
+	balanceCallbackZero          // called when priority drops to zero (positive balance exhausted)
+	balanceCallbackCount         // total number of balance callbacks
+)
+
+// BalanceTrackerSetup contains node state flags and fields used by BalanceTracker
+type BalanceTrackerSetup struct {
+	PriorityFlag, UpdateFlag, activeFlag            nodestate.Flags
+	BalanceField, negBalanceKeyField, capacityField nodestate.Field
+}
+
+// NewBalanceTrackerSetup creates a new BalanceTrackerSetup and initializes the fields
+// and flags controlled by BalanceTracker
+func NewBalanceTrackerSetup(setup *nodestate.Setup) BalanceTrackerSetup {
+	return BalanceTrackerSetup{
+		// PriorityFlag is set if the node has a positive balance
+		PriorityFlag: setup.NewFlag("priorityNode"),
+		// UpdateFlag set and then immediately reset if the balance has been updated and
+		// therefore priority is suddenly changed
+		UpdateFlag: setup.NewFlag("balanceUpdate"),
+		// BalanceField contains the NodeBalance struct which implements nodePriority,
+		// allowing on-demand priority calculation and future priority estimation
+		BalanceField: setup.NewField("balance", reflect.TypeOf(&NodeBalance{})),
+	}
+}
+
+// Init sets the fields used by BalanceTracker as an input
+func (bts *BalanceTrackerSetup) Init(negBalanceKeyField, capacityField nodestate.Field) {
+	bts.negBalanceKeyField = negBalanceKeyField
+	bts.capacityField = capacityField
+}
+
+// BalanceTracker tracks positive and negative balances for connected nodes.
+// After negBalanceKeyField is set externally, a NodeBalance is created and previous
+// balance values are loaded from the database. Both balances are exponentially expired
+// values. Costs are deducted from the positive balance if present, otherwise added to
+// the negative balance. If the capacity is non-zero then a time cost is applied
+// continuously while individual request costs are applied immediately.
+// The two balances are translated into a single priority value that also depends
+// on the actual capacity.
+type BalanceTracker struct {
+	BalanceTrackerSetup
+	clock                      mclock.Clock
+	lock                       sync.Mutex
+	ns                         *nodestate.NodeStateMachine
+	ndb                        *nodeDB
+	posExp, negExp             utils.ValueExpirer
+	posExpTC, negExpTC         uint64
+	posThreshold, negThreshold uint64
+	totalAmount                utils.ExpiredValue
```

Comment:
Here I prefer the original approach: maintain inactive and active balance.
If so, we only need to recalculate the active balance and can get rid of summed balance.

Reply:
I think you are right, will do the change soon.

Reply:
done